README.md
# Aspen
Aspen is a Discord bot


changelog.txt
**== Version v1.1.1A ==**
- [May/17/2021] - Patched On Watch chat filter
- [May/04/2021] - Patched Permissions Check on certain commands.
- [Apr/26/2021] - Fixed User Arg Parser
- [Apr/26/2021] - $whois should stop crashing if the target has no roles
**== Version v1.1.1 ==**
- [Apr/17/2021] - Commands are no longer case sensitive
- [Apr/16/2021] - Wildcard chat filter
- [Apr/15/2021] - Refactor command processor
- [Apr/14/2021] - Added more help info for commands
- [Apr/14/2021] - User argument processor function created
- [Apr/10/2021] - Command Line realtime maintenance system
**== Version v1.1.0 ==**
- [Apr/06/2021] - Rebranded
**== Version v1.0.10A ==**
- [Apr/04/2021] - Added cat and dog image command
**== Version v1.0.10 ==**
- [Mar/26/2021] - Flirt command now works via Slash Commands
- [Mar/26/2021] - Added several Slash commands
**== Version v1.0.9A ==**
- [Mar/25/2021] - Made modules reloadable even if modules crash
- [Mar/25/2021] - Ignore webhooks
- [Mar/25/2021] - Permissions management
**== Version v1.0.9 ==**
- [Mar/22/2021] - Embeds are cool! More embeds everywhere
- [Mar/21/2021] - Continued rollout of Dynamic argument parsing
- [Mar/21/2021] - Users are now ignorable
**== Version v1.0.8 ==**
- [Mar/20/2021] - Changelog is now publicly accessible
- [Mar/19/2021] - Ban and Unban commands are now functional
**== Version v1.0.7 ==**
- [Mar/16/2021] - External User ID Search command added
- [Mar/15/2021] - Avatar command added

**== Upcoming ==**
- Modules
- More intuitive Settings management system
- Converting rest of the commands to the argument system
- Consolidation of certain events into functions

**== Notes ==**
- Any changes to changelog, without version changes are
   hot-patched updates, without a bot restart.

aspen/__init__.py

from datetime import datetime
import discord
from addict import Dict
import json
import asyncio

def strfdelta(tdelta, fmt):
    d = {"days": tdelta.days}
    d["hours"], rem = divmod(tdelta.seconds, 3600)
    d["minutes"], d["seconds"] = divmod(rem, 60)
    return fmt.format(**d)

async def ping(client, message):
    elapsed_time = 0
    msgtime = message.created_at
    now = datetime.utcnow()
    #now - msgtime datetime.timedelta(0, 3, 519319)
    diff = now - msgtime
    elapsed_time = (diff.days * 86400000) + (diff.seconds * 1000) + (diff.microseconds / 1000)
    elapsed_time = str(elapsed_time)
    frameworklat = str(client.latency*1000)
    embed = discord.Embed(title="Ping",description="Bot Latency",colour=discord.Colour.from_rgb(255, 0, 242))\
        .add_field(name="Serverside", value=elapsed_time+ " ms")\
        .add_field(name="Framework", value=frameworklat[:-11]+" ms")
    print("External Chat Event at "+str(datetime.now())+" in "+str(message.channel.id))
    await message.channel.send(embed=embed)

async def selfping(client, message):
    postTime = datetime.utcnow()
    msg = await message.channel.send("####%--PING$$-"+str(postTime))
    elapsed_time = 0
    uplink_time = 0
    msgtime = msg.created_at
    now = datetime.utcnow()
    #now - msgtime datetime.timedelta(0, 3, 519319)
    diff = now - msgtime
    await msg.delete()
    elapsed_time = (diff.days * 86400000) + (diff.seconds * 1000) + (diff.microseconds / 1000)
    elapsed_time = str(elapsed_time)
    diff = msgtime-postTime
    uplink_time = (diff.days * 86400000) + (diff.seconds * 1000) + (diff.microseconds / 1000)
    uplink_time = str(uplink_time)
    frameworklat = str(client.latency*1000)
    print("External Chat Event at "+str(datetime.now())+" in "+str(message.channel.id))
    embed = discord.Embed(title="Ping", description="Self-measured Bot Latency", colour=discord.Colour.from_rgb(255, 0, 242)) \
        .add_field(name="Message Send", value=uplink_time + " ms") \
        .add_field(name="Message Receive", value=elapsed_time + " ms") \
        .add_field(name="Framework", value=frameworklat[:-11] + " ms", inline=False)
    return embed

async def ignoreUser(id, guild, state):
    if state:
        f = open("settings.json")
        settings = Dict(json.load(f))
        settings[str(guild)]["Users"]["Ignored"].setdefault(id,)
        jsonFile = open("settings.json", "w+")
        json.dump(settings, jsonFile, indent=4,)
        jsonFile.close()
    else:
        file = open("settings.json")
        settings = Dict(json.load(file))
        settings[str(guild)]["Users"]["Ignored"].pop(id)
        jsonFile = open("settings.json", "w+")
        json.dump(settings, jsonFile, indent=4, )
        jsonFile.close()

async def getChannel(client, message, type):
    f = open("settings.json")
    settings = Dict(json.load(f))
    channelid = settings[str(message.guild.id)]["Channels"][type]
    channel = client.get_channel(int(channelid))
    return channel

async def isOwner(message):
    f = open("config.json")
    config = Dict(json.load((f)))
    # await message.channel.send(str(list(config["Contributors"].values())))
    if message.author.id in list(config["Contributors"].values()):
        if "--debug" in message.content.lower():
            await message.channel.send("IsOwner=True")
        return True
    else:
        return False

async def isAdmin(message):
    return

async def isMod(message):
    return

async def userArgParse(client, message, selector):
    text = message.content.lower().split()
    replacer = text[0]
    if len(text) == 1:
        target = message.author.id
    elif text[selector].startswith('<@!') or text[selector].startswith('<@'):
        target = text[selector].replace("<@", "", 1)
        target = target.replace("!", "", 1)
        target = target.replace(">", "", 1)
    else:
        try:
            target = int(text[selector])
            if len(str(target)) > 18 and len(str(target)) < 17:
                await message.channel.send("Please ensure you used a User ID, or valid user mention.")
                return None

        except:
            await message.channel.send("Please ensure you used a User ID, or valid user mention.")
            return None
    try:
        target = await client.fetch_user(int(target))
    except:
        await message.channel.send("Could not find user!")
        return None
    try:
        target = await message.guild.fetch_member(int(target.id))
    except:
        print("Not in guild")

    return target


aspen/logevents.py
import discord
from addict import Dict
import json
from datetime import datetime
import subprocess
import aspen
try:
    f = open("settings.json")
    settings = Dict(json.load(f))
except:
    print("An error has occurred")

async def permError(client, message, permission):
    embed = discord.Embed(title="Insufficient Permissions", description="Only "+permission+" is allowed to run this command!", colour=discord.Colour.red())
    await message.channel.send(embed=embed)
async def filterLog(client, message, key):
    embed = discord.Embed(
        title="Global Filter",
        description="Banned word triggered by `Wildcard`",
        colour=discord.Colour.from_rgb(255, 0, 242)
    )
    embed.add_field(name="Banned Word", value=key)
    embed.add_field(name="Full Message", value=message.content, inline=False)
    embed.set_author(name=message.author.name + '#' + message.author.discriminator, icon_url=message.author.avatar_url)
    channel = await aspen.getChannel(client, message, "Logging")
    print("Internal Chat Event at " + str(datetime.now()) + " in " + str(channel.id))
    await channel.send(embed=embed)
    return
async def internalEventLog(client, guild, user, action, string=None):
    f = open("settings.json")
    settings = Dict(json.load(f))
    if string is None:
        string = "\u200b"
    embed = discord.Embed(title=action,description=string)
    embed.set_author(name=user.name+"#"+user.discriminator, icon_url=user.avatar_url)
    channelid = settings[str(guild)]["Channels"]["Logging"]
    channel = client.get_channel(int(channelid))
    await channel.send(embed=embed)
    return
async def deleteLog(client, message):
    if message.author == client.user:
        return

    f = open("settings.json")
    settings = Dict(json.load(f))
    embed = discord.Embed(colour=discord.Colour.red())
    try:
        targetLoc = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")+"_"+message.attachments[0].filename
        action = "sudo wget "+message.attachments[0].proxy_url+" -O /var/www/file-share/1/discord-del/"+targetLoc
        print(subprocess.check_output(action, shell=True))
        imagePresent = True

    except:
        imagePresent = False
    if imagePresent == True:

        embed.set_image(url="https://deleted.hypr.ax/"+targetLoc)
        embed.add_field(name="Location", value="[File](https://deleted.hypr.ax/" + targetLoc, inline=False)
        # print(targetLoc)

    embed.add_field(name="\u200b", value="**Message sent by <@"+str(message.author.id)+"> deleted in <#"+str(message.channel.id)+">** \n"+str(message.content))
    embed.set_author(name=message.author.name+'#'+message.author.discriminator, icon_url=message.author.avatar_url)
    embed.set_footer(text="Author ID: "+str(message.author.id)+" Message ID: "+str(message.id)+" Timestamp: "+str(datetime.utcnow()))
    channel = settings[str(message.guild.id)]["Channels"]["Logging"]
    await client.get_channel(int(channel)).send(embed=embed)

async def settingsLog(client, message, action):
    user = message.author
    guild = message.guild.id
    f = open("settings.json")
    settings = Dict(json.load(f))
    embed = discord.Embed(
        title="Settings Modification",
        description=user.mention+" has "+action,
        colour=discord.Colour.green()
    )
    embed.set_author(name=user.name+"#"+user.discriminator, icon_url=user.avatar_url)
    channel = settings[str(guild)]["Channels"]["Logging"]
    await client.get_channel(int(channel)).send(embed=embed)
    return

aspen/messaging.py
import aspen

async def send(client, message):
    await message.delete()
    if message.author.guild_permissions.administrator == True or aspen.isOwner(message):
        text = message.content.split()
        channel = client.get_channel(int(text[1]))
        textTarg = str(text[0]+" "+text[1])
        text = message.content.replace(textTarg, "", 1)
        await channel.send(text)
        return
    return
async def targetSend(client, message, channelid):
    if message.author.guild_permissions.administrator == True or aspen.isOwner(message):
        text = message.content
        text = text.replace(message.content.split()[0]+" ", "", 1)
        print(text)
        channel = client.get_channel(channelid)
        await message.delete()
        await channel.send(text)
        return
    return


aspen/moderation.py
import aspen
import discord
from aspen import logevents

async def purge(client, message):
    if message.author.guild_permissions.administrator == True or aspen.isOwner(message):
        text = message.content.split()
        if len(text) != 2:
            await message.channel.send(embed=discord.Embed(title="Purge", description="Purges messages from the channel",
                                                               colour=discord.Colour.from_rgb(255, 0, 242))
                                           .add_field(name="Permissions", value="Administrator")
                                           .add_field(name="Usage", value="$purge <# of messages>"))
            return
        await message.delete()
        await message.channel.purge(limit=int(text[1]))
        stringy = text+" Messages deleted in <#"+str(message.channel.id)+">"
        await logevents.internalEventLog(client, message.guild.id, message.author, "Purge", stringy)

async def delete(client, message):
    if message.author.guild_permissions.administrator or aspen.isOwner(message):
        text = message.content.split()
        ctx = text[1]
        target = text[2]
        ctx = client.get_channel(int(ctx))
        msg = ctx.fetch_message(int(target))
        await msg.delete()
        return

async def warn(client, message):
    return



aspen/extension.py
import discord
import aspen
from aspen import  moderation, logevents
from datetime import datetime
import resource
import requests
import asyncio
from discord_components import DiscordComponents, Button, ButtonStyle, InteractionType

async def init(client, ddb, message, command, debugstate):
    if message.guild != None:
        member = await message.guild.fetch_member(message.author.id)
        if not member.guild_permissions.manage_messages:
            for x in ["discord.gg/", "discord.com/invite", "discordapp.com/invite", "watchanimeattheoffice.com/invite", "discord.co/invite"]:
                if x in message.content.lower():
                    await message.delete()
                    key = "Discord Invite Link"
                    await logevents.filterLog(client, message, key)
                    return

    if message.author == client.user:
        return
    if message.author.bot:
        return

    if command == "!call" and message.guild.id == 735872336045277234:
        await client.get_channel(758729070111490109).send("@everyone Call from: "+message.author.name+"#"+message.author.discriminator+" at "+str(message.jump_url))
        await message.channel.send("Moderator call. Please standby.")
    for x in message.mentions:
        if x == client.user:
            embed = discord.Embed(
                title="Aspen#8530",
                description="""A bot developed and maintained by <@193112730943750144>

    Currently running **Aspen v1.1.1A**

                        """
            )
            embed.set_footer(text="Message will delete after 30 seconds")
            if message.guild != None:
                embed.add_field(name="Shard ID", value=str(message.guild.shard_id))
            embed.set_thumbnail(
                url="https://cdn.discordapp.com/avatars/371491028873773079/86216d3ef0a07016e5f287d6de7953ad.png")
            # print("Internal Chat Event at " + str(datetime.now()) + " in " + str(message.channel.id))
            msg = await message.channel.send(embed=embed, delete_after=30, components=[Button(style=ButtonStyle.red, label="Action")])
            res = await ddb.wait_for_interact("button_click")
            if res.channel == msg.channel:
                await res.respond(
                    type=InteractionType.ChannelMessageWithSource,
                    content=f'{res.component.label} clicked'
                )
            await msg.edit(embed=embed, components=[Button(style=ButtonStyle.red, label="Clicked")])
    return


async def ext(client, message, command, debugstate):
    text = message.content.split()
    if "--debug" in message.content:
        await message.channel.send("Extensions Command: "+command)
    if command == "$ping":
        await aspen.ping(client, message)
    if command == "$av":
        if len(text) == 2:
            if text[1] == "help":
                await message.channel.send(
                embed=discord.Embed(title="Avatar", description="Displays the Avatar of a selected user",
                                    colour=discord.Colour.from_rgb(255, 0, 242))
                .add_field(name="Permissions", value="Everyone")
                .add_field(name="Usage", value="$av <User>"))
                return
        target = await aspen.userArgParse(client,  message, 1)
        if target is None:
            return
        avatar_loc = target.avatar_url
        embed = discord.Embed(title="Avatar for " + target.name + "#" + target.discriminator,
                              colour=discord.Colour.from_rgb(255, 0, 242))
        embed.set_image(url=avatar_loc)
        embed.set_author(name=message.author.name + "#" + message.author.discriminator,
                         icon_url=message.author.avatar_url)
        await message.channel.send(embed=embed)
    if command == "$s":
        if message.author.id == 193112730943750144:
            text = message.content.split()
            if len(text) != 2:
                await message.channel.send("Invalid syntax")
                return
            r = requests.post("https://s.hypr.ax/shorten/"+text[1], headers={"Authorization":"DeyhX6AFpkV7uL93L4u6dCyaSbiyfFB4iQmLvhkx"})
            if r.status_code==200:
                await message.channel.send(r.text)
            else:
                await message.channel.send("Invalid URL")
                return
    if command == "$invite":
        if message.author.guild_permissions.administrator == True:
            link = await message.channel.create_invite(max_age=300)
            await message.channel.send("Here is an instant invite to your server: " + str(link))
    if command == '$cdel':
        text = message.content.split()
        await message.delete()
        if message.author.id != 193112730943750144:
            return
        channel = client.get_channel(text[1])
        msg = channel.fetch_message(text[2])
        await msg.delete()
    if command == '$info':
        usage = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss/1000
        await message.channel.send(str(usage/1000))
    if message.content.startswith('$changelog'):
        file1 = open("changelog.txt")
        log = ""
        for line in file1:
            log = log+" "+line
        embed = discord.Embed(title="Changelog 🛠", description=log, colour=discord.Colour.from_rgb(255,0,242))\
            .set_footer(text="Use $help for commands")
        await message.channel.send(embed=embed)
    if command == '$ban':
        if message.author.guild_permissions.ban_members or await aspen.isOwner(message):
            if len(text) == 1 or (len(text) == 2 and text[1] == "help"):
                await message.channel.send(embed=discord.Embed(title="Ban", description="Bans a member from the Server",
                                                               colour=discord.Colour.from_rgb(255, 0, 242))
                                           .add_field(name="Permissions", value="Ban Members")
                                           .add_field(name="Usage", value="$ban <User>"))
                return "Exit"



            user = await aspen.userArgParse(client,  message, 1)
            if message.guild.get_member(user.id) is not None:
                if not await aspen.isOwner(message):
                    if message.author.top_role <= user.top_role:
                        await message.channel.send("You do not meet the permission requirements to assign that role!")
                        return
            if user is None:
                return "Exit"
            if user.id == 193112730943750144 and await aspen.isOwner(message):
                if "--force" in message.content.lower():
                    print("banning bot owner")
                else:
                    await message.channel.send("You may not ban the **Bot Owner**")
                    return "Exit"
            await message.guild.ban(user=user,delete_message_days=0)
            await message.channel.send(embed=discord.Embed(title="User Banned!",description="Banned "+user.name+"#"+user.discriminator,colour=discord.Colour.red()))
    if command == '$cban':
        if aspen.isOwner(message):
            if "--force" in message.content:
                message.content = message.content.replace("--force", "", 1)
            if len(text) == 1:
                await message.channel.send(embed=discord.Embed(title="Contextual Ban", description="Bans a member from a specified guild",
                                                               colour=discord.Colour.from_rgb(255, 0, 242))
                                           .add_field(name="Permissions", value="Bot Owner")
                                           .add_field(name="Usage", value="$cban <Guild ID> <User ID>"))
                return
            user = await aspen.userArgParse(client,  message, 2)
            if user is None:
                return
            if user.id == 193112730943750144:
                await message.channel.send("You may not ban the **Bot Owner**")
                return
            guild = client.get_guild(int(text[1]))
            await guild.ban(user=user,delete_message_days=0)
            await message.channel.send(embed=discord.Embed(title="User Banned!",description="Banned "+user.name+"#"+user.discriminator,colour=discord.Colour.red()))
    if command == '$cunban':
        if aspen.isOwner(message):
            if len(text) == 1:
                await message.channel.send(embed=discord.Embed(title="Contextual Unan", description="Unans a member from a specified guild",
                                                               colour=discord.Colour.from_rgb(255, 0, 242))
                                           .add_field(name="Permissions", value="Bot Owner")
                                           .add_field(name="Usage", value="$cunban <Guild ID> <User ID>"))
                return
            user = await aspen.userArgParse(client,  message, 2)
            if user is None:
                return
            guild = client.get_guild(int(text[1]))
            await guild.unban(user)
            await message.channel.send(embed=discord.Embed(title="User Unanned!",description="Unanned "+user.name+"#"+user.discriminator))
        else:
            return
    if command == '$unban':
        if message.author.guild_permissions.ban_members:
            if len(text) == 1:
                await message.channel.send(
                    embed=discord.Embed(title="Unban", description="Unbans a user from this guild",
                                        colour=discord.Colour.from_rgb(255, 0, 242))
                    .add_field(name="Permissions", value="Ban Members")
                    .add_field(name="Usage", value="$unban <User ID>"))
                return
            user = await aspen.userArgParse(client,  message, 1)
            if user is None:
                return
            await message.guild.unban(user)
            await message.channel.send("Unbanned "+user.name+"#"+user.discriminator)
            #
            #    await message.channel.send("User either not banned or not found")
    if command == "$ofd":
        await message.delete()
        if message.author.id == 193112730943750144 or message.author.id == 219838416878174209:
            text = message.content.split()
            channel = client.get_channel(736432247842013234)
            msg = await channel.fetch_message(text[1])
            await msg.delete()
        else:
            return
    if command == '$test':
        #await message.channel.send("done")
        return "exit"
    if command == '$echo':
        if aspen.isOwner(message):
            await message.channel.send(message.content)
    if message.author.id == 750835145556230206:
        for x in [":smiling_face_with_3_hearts:", "babe", "🥰"]:
            if x in message.content.lower():
                await message.delete()
    if command == '$cat':
        msg = await message.channel.send("Searching for a cat...")
        r = requests.get("http://aws.random.cat/meow")
        file = r.json()['file']
        embed = discord.Embed(title="Found a Cat! 😺", colour=discord.Colour.from_rgb(255,0,242))\
            .set_image(url=file)
        await msg.edit(embed=embed, content="\u2000")
        return "exit"
    if command == '$dog':
        msg = await message.channel.send("Searching for a dog...")
        r = requests.get("https://random.dog/woof.json")
        file = r.json()
        while int(file['fileSizeBytes']) >= 100000:
            #print("gettings")
            r = requests.get("https://random.dog/woof.json")
            file = r.json()
        embed=discord.Embed(title="Found a Dog! 🐕", colour=discord.Colour.from_rgb(255,0,242))\
            .set_image(url=file['url'])
        await msg.edit(embed=embed, content="\u2000")
    if command == '$btest':
        if message.author.id != 19311273094375014:
            print("ok")
        else:
            return
        text = message.content.split()
        await message.delete()
        if len(text) < 3:
            await message.channel.send("Incorrect!", delete_after=1)
        guild = client.get_guild(735872336045277234)
        try:
            user = await guild.fetch_member(int(text[1]))
        except:
            user = await client.fetch_user(int(text[1]))
        if user is None:
            user = await client.fetch_user(int(text[1]))
        if user is None:
            await message.channel.send("Invalid!!", delete_after=1)
        pfp = str(user.avatar_url)
        try:
            username = str(user.nick)
        except:
            username = str(user.name)
        if username == "None":
            username = str(user.name)
        if user.id == 193112730943750144:
            username = "TheНyperdrive"
        content = message.content.replace(text[0] + " " + text[1], "", 1)

        r = requests.post(
            "https://discord.com/api/webhooks/825258566310297620/GCEhYEtWMW_WH2HMZv4VSaALspiSa3Sc8M3hjnhGlqh-l7TBa1BbbFPL45HG2Xb1DHIG",
            headers={'User-Agent': 'Mozilla/5.0'},
            data={'content': content, 'username': username, 'avatar_url': pfp}
            )
        # #print(r.text)
    if command == '$stest':
        if message.author.id == 193112730943750144 or message.author.id == 750835145556230206 or message.author.id == 219838416878174209:
            print("ok")
        else:
            return
        text = message.content.split()
        await message.delete()
        if len(text) < 3:
            await message.channel.send("Incorrect!", delete_after=1)
        guild = client.get_guild(806151212814565386)
        try:
            user = await guild.fetch_member(int(text[1]))
        except:
            user = await client.fetch_user(int(text[1]))
        if user is None:
            user = await client.fetch_user(int(text[1]))
        if user is None:
            await message.channel.send("Invalid!!", delete_after=1)
        pfp = str(user.avatar_url)
        try:
            username = str(user.nick)
        except:
            username = str(user.name)
        if username == "None":
            username = str(user.name)
        if user.id == 193112730943750144:
            username = "TheНyperdrive"
        content = message.content.replace(text[0] + " " + text[1], "", 1)

        r = requests.post(
            "https://discord.com/api/webhooks/825097257107718204/F2Y-WLDChanW3YvA1GVVrp4zv8A_dBnQwM6yt8OxWyzE2GheIy0aWIkIMQQZoD2xyyc2",
            headers={'User-Agent': 'Mozilla/5.0'},
            data={'content': content, 'username': username, 'avatar_url': pfp}
            )
        # #print(r.text)
    if command == '$msend':
        if message.author.id == 193112730943750144 or message.author.id == 373992293071585281:
            print("ok")
        else:
            return
        text = message.content.split()
        await message.delete()
        if len(text) < 3:
            await message.channel.send("Incorrect!", delete_after=1)
        guild = client.get_guild(615608144726589457)
        try:
            user = await guild.fetch_member(int(text[1]))
        except:
            user = await client.fetch_user(int(text[1]))
        if user is None:
            user = await client.fetch_user(int(text[1]))
        if user is None:
            await message.channel.send("Invalid!!", delete_after=1)
        pfp = str(user.avatar_url)
        try:
            username = str(user.nick)
        except:
            username = str(user.name)
        if username == "None":
            username = str(user.name)
        content = message.content.replace(text[0] + " " + text[1], "", 1)

        r = requests.post(
            "https://discord.com/api/webhooks/826588669074210836/uXOx6HCgN8QH3hS2s256-ixYyb7GjBZMTMtEKs0UeSvUu-L_4HU8Ny5HITNcthy6yl89",
            headers={'User-Agent': 'Mozilla/5.0'},
            data={'content': content, 'username': username, 'avatar_url': pfp}
            )

    if command == '$cpurge':
        channel = client.get_channel(int(text[1]))
        await channel.purge(limit=int(text[2]))
    if command == '$ctest':
        if message.author.id == 193112730943750144:
            print("ok")
        else:
            return
        text = message.content.split()
        await message.delete()
        if len(text) < 3:
            await message.channel.send("Incorrect!", delete_after=1)
            return
        guild = client.get_guild(603203154091311104)
        try:
            user = await guild.fetch_member(int(text[1]))
        except:
            user = await client.fetch_user(int(text[1]))
        if user is None:
            user = await client.fetch_user(int(text[1]))
        if user is None:
            await message.channel.send("Invalid!!", delete_after=1)
        pfp = str(user.avatar_url)
        try:
            username = str(user.nick)
        except:
            username = str(user.name)
        if username == "None":
            username = str(user.name)
        content = message.content.replace(text[0] + " " + text[1], "", 1)

        requests.post(
            "https://discord.com/api/webhooks/748793443177857035/8lSRw-21zGJfCO21D8BKZJjiC-R16pUBUIsYu9M6KZy0vEFlwnkLuTiFE3yMEljEUwxu",
            headers={'User-Agent': 'Mozilla/5.0'},
            data={'content': content, 'username': username, 'avatar_url': pfp}
        )
            # #print(r.text)
        if command == '$msend':
            if aspen.isOwner(message) or message.author.id == 373992293071585281:
                print("ok")
            else:
                return
            text = message.content.split()
            await message.delete()
            if len(text) < 3:
                await message.channel.send("Incorrect!", delete_after=1)
            guild = client.get_guild(615608144726589457)
            try:
                user = await guild.fetch_member(int(text[1]))
            except:
                user = await client.fetch_user(int(text[1]))
            if user is None:
                user = await client.fetch_user(int(text[1]))
            if user is None:
                await message.channel.send("Invalid!!", delete_after=1)
            pfp = str(user.avatar_url)
            try:
                username = str(user.nick)
            except:
                username = str(user.name)
            if username == "None":
                username = str(user.name)
            content = message.content.replace(text[0] + " " + text[1], "", 1)

            r = requests.post(
                "https://discord.com/api/webhooks/826588669074210836/uXOx6HCgN8QH3hS2s256-ixYyb7GjBZMTMtEKs0UeSvUu-L_4HU8Ny5HITNcthy6yl89",
                headers={'User-Agent': 'Mozilla/5.0'},
                data={'content': content, 'username': username, 'avatar_url': pfp}
            )
        # #print(r.text)
    if command == '$timein':
        text = message.content.split()
        if len(text) != 2:
            await message.channel.send(embed=discord.Embed(title=""))
    if command == '$roleiter':
        if message.author.guild_permissions.manage_roles:
            string_length = len(str(message.guild.roles))
            if string_length > 1900:
                chunklength = 1899
                chunks = [str(message.guild.roles)[i:i + chunklength] for i in range(0, len(str(message.guild.roles)), chunklength)]
                starter = 1
                for chunk in chunks:
                    # embedVar = discord.Embed(title="Linux Command Executor",
                    #        description=chunk,
                    #        color=0x9CAFBE,
                    #        inline=True)
                    if starter == 1:
                        await message.channel.send(content=chunk)
                        starter = 2
                    else:
                        await message.channel.send(chunk)
            else:
                # embedVar = discord.Embed(title="Linux Command Executor",
                #            description=result,
                #            color=0x9CAFBE,
                #            inline=True)
                await message.channel.send(content=str(message.guild.roles))
    if command == '$extid':
        try:
            target = await client.fetch_user(int(text[1]))
        except:
            await  message.channel.send("Could not find user")
            return
        embed=discord.Embed(title="External ID Search",
                            description=text[1],
                            colour=discord.Colour.from_rgb(255, 0, 242)
                            )\
            .add_field(name="User", value=target.name+"#"+target.discriminator, inline=False)\
            .add_field(name="Registered:", value=str(target.created_at)[:-7])\
            .set_thumbnail(url=target.avatar_url)\
            .set_footer(text=message.author.name+"#"+message.author.discriminator+" | "+datetime.now().strftime("%b %d, %Y %H:%M:%S"), icon_url=message.author.avatar_url)
        if target.bot:
            embed.add_field(name="Bot User", value=":white_check_mark: True")
        else:
            embed.add_field(name="Bot User", value=":x: False")
        await message.channel.send(embed=embed)
        return "exit"
    return



